verifier/src/lib.rs

```diff
@@ -202,17 +203,35 @@ where
         // when the trace contains auxiliary segments, append auxiliary trace elements at the
         // end of main trace elements for both current and next rows in the frame. this is
         // needed to be consistent with how the prover writes OOD frame into the channel.
-
         let mut current = ood_main_trace_frame.current().to_vec();
         current.extend_from_slice(aux_trace_frame.current());
-        public_coin.reseed(H::hash_elements(&current));
 
         let mut next = ood_main_trace_frame.next().to_vec();
         next.extend_from_slice(aux_trace_frame.next());
-        public_coin.reseed(H::hash_elements(&next));
+
+        // merge `current` and `next` in an interleaved way in order to obtain a single vector.
+        let trace_states = vec![current, next];
+        let width = trace_states[0].len();
+        let mut result = vec![E::ZERO; width * FRAME_SIZE];
+        for i in 0..width {
+            for j in 0..FRAME_SIZE {
+                result[i * FRAME_SIZE + j] = trace_states[j][i];
+            }
+        }
+        public_coin.reseed(H::hash_elements(&result));
     } else {
-        public_coin.reseed(H::hash_elements(ood_main_trace_frame.current()));
-        public_coin.reseed(H::hash_elements(ood_main_trace_frame.next()));
+        let trace_states = vec![
+            ood_main_trace_frame.current().to_vec(),
+            ood_main_trace_frame.next().to_vec(),
+        ];
+        let width = trace_states[0].len();
+        let mut result = vec![E::ZERO; width * FRAME_SIZE];
+        for i in 0..width {
+            for j in 0..FRAME_SIZE {
+                result[i * FRAME_SIZE + j] = trace_states[j][i];
+            }
+        }
+        public_coin.reseed(H::hash_elements(&result));
     }
```

Comment:
I wonder if this could be somehow done inside `OodFrame` struct. I don't have a great suggestion how to do this yet, but I'll try to come up with something.

Reply:
Agree, all options I could come up with are either a bit "hacky" or require significant changes to some of the structs.

Reply:
One potential solution (though still not ideal) is to put this logic into [TraceOodFrame](https://github.com/facebook/winterfell/blob/main/verifier/src/channel.rs#L346) struct. Basically, we could change this strict to look something like:

```Rust
pub struct TraceOodFrame<E: FieldElement> {
    values: Vec<E>,
    main_trace_width: usize,
    aux_trace_width: usize,
}

impl<E: FieldElement> TraceOodFrame<E> {
    pub fn new(values: Vec<E>, main_trace_width: usize, aux_trace_width: usize) -> Self {
        Self {
            values,
            main_trace_width,
            aux_trace_width,
        }
    }

    pub fn values(&self) -> &[E] {
        &self.values
    }

    pub fn main_frame(&self) -> EvaluationFrame<E> {
        todo!("build the main frame here from relevant parts of self.values")
    }

    pub fn aux_frame(&self) -> Option<EvaluationFrame<E>> {
        if self.aux_trace_width == 0 {
            None
        } else {
            todo!("build the aux frame here from relevant parts of self.values")        
        }
    }
}
```

Then, we'd update [read_ood_trace_frame()](https://github.com/facebook/winterfell/blob/main/verifier/src/channel.rs#L143) method to look as follows:
```Rust
pub fn read_ood_trace_frame(&mut self) -> TraceOodFrame {
    let frame = self.ood_trace_frame.take().expect("already read")
}
```

And finally, we'd update [parse()](https://github.com/facebook/winterfell/blob/main/air/src/proof/ood_frame.rs#L88) method of `OodFrame` to return just two vectors: one with all trace values and the other one with constraint evaluations. The first vector would then be passed to `TraceOodFrame` as a construction parameters.

We still have the logic split between two places - but I feel it would be a bit better encapsulated.